opm/simulators/flow/FlowProblemComp.hpp

```diff
@@ -438,13 +449,43 @@ class FlowProblemComp : public FlowProblem<TypeTag>
                     dofFluidState.setPressure(phaseIdx, pressure);
             }
 
-            for (unsigned compIdx = 0; compIdx < numComponents; ++compIdx) {
-                const std::size_t data_idx = compIdx * numDof + dofIdx;
-                const Scalar xmf = xmfData[data_idx];
-                const Scalar ymf = ymfData[data_idx];
+            if (has_xmf && has_ymf) {
+                const auto& xmfData = fp.get_double("XMF");
+                const auto& ymfData = fp.get_double("YMF");
+                for (unsigned compIdx = 0; compIdx < numComponents; ++compIdx) {
+                    const std::size_t data_idx = compIdx * numDof + dofIdx;
+                    const Scalar xmf = xmfData[data_idx];
+                    const Scalar ymf = ymfData[data_idx];
+
+                    dofFluidState.setMoleFraction(FluidSystem::oilPhaseIdx, compIdx, xmf);
+                    dofFluidState.setMoleFraction(FluidSystem::gasPhaseIdx, compIdx, ymf);
+                }
+            }
 
-                dofFluidState.setMoleFraction(FluidSystem::oilPhaseIdx, compIdx, xmf);
-                dofFluidState.setMoleFraction(FluidSystem::gasPhaseIdx, compIdx, ymf);
+            if (has_zmf) {
+                const auto& zmfData = fp.get_double("ZMF");
+                for (unsigned compIdx = 0; compIdx < numComponents; ++compIdx) {
+                    const std::size_t data_idx = compIdx * numDof + dofIdx;
+                    const Scalar zmf = zmfData[data_idx];
+                    // TODO: I know we do this for co2_ptflash example, but maybe we should not do it this way?
+                    if (gas_active) {
+                        const auto& sat_gas = dofFluidState.saturation(gasPhaseIdx);
+                        if (sat_gas > 0.) {
+                            dofFluidState.setMoleFraction(FluidSystem::gasPhaseIdx, compIdx, zmf);
+                        } else {
+                            dofFluidState.setMoleFraction(FluidSystem::gasPhaseIdx, compIdx, 0.);
+                        }
```

Comment:
An alternative way to write this is as
```C++
const auto ymf = (dofFluidState.saturation(gasPhaseIdx) > 0) ? zmf : Scalar{0};
dofFluidState.setMoleFraction(FluidSystem::gasPhaseIdx, compIdx, ymf);
```
which may or may not be easier to read.

Reply:
done.